keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,468 @@
+
+#### Could it done on the client side instead of inside etcd? 
+One could argue that it is easy to know if a feature enabled for the whole cluster by querying each member if the feature is enabled on the client side. But there are several caveates with this approach:
+* not all members are available all the time. if some member is not available, it is questionable if the feature can be used, and that could break the HA assumption. 
+* some feature might change how a raft message is sent and applied, and the order of messages relative to the index the feature is enabled is critical. Here the consistent index is more important than the time, and it is hard to get that outside etcd. 
+
+#### Should we prevent a new member with an inconsistent feature gate from joining the cluster?
+
+One way to prevent inconsistencies of feature gate between different members is to prevent a new member with an inconsistent feature gate from joining the cluster in the first place. 
+We do not think this is the right way because we want to keep the ability to turn on/off a feature in a rolling sequence without bringing down the whole cluster. 
+
+
+### Risks and Mitigations
+
+#### Backport Risks
+
+Because we are adding a new type of raft request, if we want to support feature gate as soon as in 3.6, we have to backport the proto changes and add no-op apply logic in 3.5. 
+There is the potential risk of changing a stable release. But we do not think this is a real risk, because it should be a no-op change in 3.5. 
+When upgrading etcd cluster from 3.5 to 3.6, the users would need to upgrade to the latest 3.5 release first.
+
+If we do not backport to 3.5, we would need to add no-op changes in 3.6, and wait till 3.7 to add the real feature gate capabilities, which could be a long cycle of several years.
+
+#### Data Compatibility Risks during Feature Value Change
+
+There are some features that would need to touch data in the apply/storage layer. Disabling them might result in orphaned data or unrecognizable data that could be breaking.
+How do we handle the data change? 
+1. if the etcd server knows this feature, it should be able to handle the data when the feature is on or off. How it handles the data at WAL entry `i` is determined by the cluster feature flag set from raft before WAL entry `i`. 
+1. if the etcd server does not know this feature, i.e. its version is not compatible with this feature, it means the data is not compatible with the storage version. The storage version migration should handle cleaning up older or newly introduced data.
+
+Overall, when developers introduce new data with a feature, they should be careful to include data cleaning in `schema.UnsafeMigrate`. But we prefer not to have a mechanism to disallow such features to be turned off after they are turned on because new features could be buggy, we need to keep the option to be able to turn it off without bring down the whole cluster.
+
+#### Feature Implementation Change Risks
```

Comment:
Can you cover a broader issue of  how how the feature gates will work with downgrade/upgrade? What happens if cluster is being downgraded to version that doesn't support a feature gate? The solution should be the same, using Cluster Version.

Reply:
added all the scenarios.